### PR TITLE
Add local reply owner information in filter state

### LINKED ISSUE #28834: Add into dynamic metadata the filter name that generate a local replay
Add into dynamic metadata the filter name that generate a local replay

It could be useful lo allow http_connection_manager add metadata for the filter name that stop iteration and generate a local replay
That metadata could be consumed by access_log or eventually other custom filter

### PR DESCRIPTION
**Pull request by juanmolle**
Commit Message:  Add local reply owner information in filter state
Additional Description: This PR is related to #28834 Adding local reply owner information in filter_state to be consume by access log or any custom filter.
Risk Level: Low
Testing: UT
Docs Changes: Yes
Release Notes: Yes
Platform Specific Features: n/a
Fixes #28834

### CODE REVIEW

**Comment by juanmolle** [comment]
I would like some feedback if this approach is the best, I was considering to add this info as attribute in the streamInfo, like CodeDetails. any recomendation?

**Comment by juanmolle** [comment]
@ramaraochavali

File: source/common/http/filter_manager.cc
Comment (wbpcode) [inline source/common/http/filter_manager.cc:465]: Basically, we keep the `filter_name` here just for backward compatibility and it maybe removed from the filter context in the future. So, it's better to use `config_name` only here.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for the contribution. First comment to start the review.

File: source/common/http/filter_manager.cc
Comment (juanmolle) [reply source/common/http/filter_manager.cc:465]: how can I know the extension type? details has some meaning depending the extension, for example ext_authz, it may contains server error, or token invalid

Comment (wbpcode) [reply source/common/http/filter_manager.cc]: Basically, the config_name should be unique in the filter chain. So, the config_name should be enough to tell the related filter type info.

**Comment by juanmolle** [comment]
I set this ti draft because I found 1 scenario not working as expected, in the case the filter is a composite the config name is for the composite, but it should be the inner filter in the composite. , in the following filter I get level-00001 instead of EXPECTED-FILTER-NAME

@wbpcode any hit you can provide me?

                  name: level-00001
                  typed_config:
                    "@type": type.googleapis.com/envoy.extensions.common.matching.v3.ExtensionWithMatcher
                    matcher:
                      matcher_list:
                        matchers:
                        - predicate:
                            single_predicate:
                              input:
                                name: request-headers
                                typed_config:
                                  "@type": type.googleapis.com/envoy.type.matcher.v3.HttpRequestHeaderMatchInput
                                  header_name: ":path"
                              value_match:
                                prefix: "/api/httpbin/"
                          on_match:
                            action:
                              name: execute-filter-action
                              typed_config:
                                "@type": type.googleapis.com/envoy.extensions.filters.http.composite.v3.ExecuteFilterAction
                                typed_config:
                                  name: EXPECTED-FILTER-NAME
                                  typed_config:
                                    "@type": type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthz
                                    grpc_service:
                                      envoy_grpc:
                                        cluster_name: cluster-name
                                      timeout: 5s
                                    transport_api_version: V3
                    extension_config:
                      name: composite
                      typed_config:
                        "@type": type.googleapis.com/envoy.extensions.filters.http.composite.v3.Composite

**Comment by juanmolle** [comment]
After seeing this, the behavior I'm seeing I think is correct, the filter config_name correspond to the composite filter name.
the problem is that the typed_config name for the action in the composite get lost, and probably the info I'm looking for should be publish by the composite extension. Make sense?

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks for you contribution. I think a release note is necessary.

cc @kyessenov Could you also take a look at this when you have free time? I think you have lots of great exp on this area.

File: source/common/http/filter_manager.h
Comment (juanmolle) [reply source/common/http/filter_manager.h]: I will need help in docs, my english is not good

File: docs/root/configuration/advanced/well_known_filter_state.rst
Comment (wbpcode) [inline docs/root/configuration/advanced/well_known_filter_state.rst:59]: `envoy.filters.network.http_connection_manager.local_reply_owner`

This filter state doesn't belong to one specific filter, so may this name is better.

**Review by wbpcode — CHANGES_REQUESTED** [review]
Thanks, a minor comment to the naming.

File: docs/root/configuration/advanced/well_known_filter_state.rst
Comment (wbpcode) [inline docs/root/configuration/advanced/well_known_filter_state.rst:61]: Shared filter status for logging which filter in the HTTP filter chain sent the local reply.

File: docs/root/configuration/advanced/well_known_filter_state.rst
Comment (wbpcode) [inline docs/root/configuration/advanced/well_known_filter_state.rst:63]: ``filter_config_name``: filter configuration name in the HTTP filter chain;

File: source/common/http/filter_manager.h
Comment (wbpcode) [inline source/common/http/filter_manager.h:60]: Hmmm, I doubt if this detail is necessary here? I think we have record this detail in the stream info?

Comment (juanmolle) [reply source/common/http/filter_manager.h:60]: yes, it is available as attribute

Comment (wbpcode) [reply source/common/http/filter_manager.h:61]: If wasm filter connot comsume this value, then it would be better to enhance the ABI to expose it. So, just let's remove it here.

**Comment by wbpcode** [comment]
Could you also check the coverage CI? Seems the failure is related to this patch. I think you need to ensure all new code you added should be covered.

**Comment by juanmolle** [comment]
> Could you also check the coverage CI? Seems the failure is related to this patch. I think you need to ensure all new code you added should be covered.

Yes, I see now proto serialization is not covered, will add a test for that and remove details.

File: changelogs/current.yaml
Comment (wbpcode) [inline changelogs/current.yaml:22]:
```
- area: filter state
  change: |
    Added config name of filter that invoking local reply in filter state with key 
    ``envoy.filters.network.http_connection_manager.local_reply_owner``.
    See :ref:`the well-known filter state keys <well_known_filter_state>`. for more detail.
```

File: docs/root/configuration/advanced/well_known_filter_state.rst
Comment (wbpcode) [inline docs/root/configuration/advanced/well_known_filter_state.rst:64]: please update the docs also.

**Review by wbpcode — COMMENTED** [review]
Thanks. LGTM overall, except the release note and docs comments. And please check all possible spelling error.

Will defer this to @kyessenov to take a look again. It's expert in the area of filter state usage.

File: changelogs/current.yaml
Comment (kyessenov) [inline changelogs/current.yaml:22]: filter that invoking local reply -> filter sending a local reply

File: changelogs/current.yaml
Comment (kyessenov) [inline changelogs/current.yaml:24]: Spurious dot before `for`.

File: source/common/http/filter_manager.cc
Comment (kyessenov) [inline source/common/http/filter_manager.cc:468]: You can drop `this->`, and I think this should be readonly?

File: source/common/http/filter_manager.cc
Comment (kyessenov) [inline source/common/http/filter_manager.cc:1727]: Can you combine the two implementations into the base class maybe?

File: source/common/http/filter_manager.h
Comment (kyessenov) [inline source/common/http/filter_manager.h:39]: Style nit: LocalReplyFilterStateKey

File: source/common/http/filter_manager.h
Comment (kyessenov) [inline source/common/http/filter_manager.h:41]: This is a filter state object, so maybe call it LocalReplyOwnerObject?

File: docs/root/configuration/advanced/well_known_filter_state.rst
Comment (kyessenov) [inline docs/root/configuration/advanced/well_known_filter_state.rst:62]: I don't think you implemented the fields, and we probably don't need it for now. Maybe just say this usually serializes to the filter config name?

File: source/common/http/filter_manager.h
Comment (kyessenov) [inline source/common/http/filter_manager.h:49]: Let's keep it simple and simply use a string value as serialized, the same as string serialization.

**Review by kyessenov — CHANGES_REQUESTED** [review]

**Comment by kyessenov** [comment]
/wait-any
Please merge main as well.

File: source/common/http/filter_manager.h
Comment (juanmolle) [reply source/common/http/filter_manager.h:49]: you mean to keep the struct but with a string instead of map?

Comment (juanmolle) [reply source/common/http/filter_manager.h:49]: Are we sure in the future this will not include other value?

Comment (kyessenov) [reply source/common/http/filter_manager.h]: I'm not sure we won't need it, but that's OK since this is primarily for telemetry AFAICT, and is fairly free-form.

File: source/common/http/filter_manager.cc
Comment (kyessenov) [inline source/common/http/filter_manager.cc:485]: Sorry if I wasn't clear - but these two lines could be added as a method in `ActiveStreamFilterBase` that is then called from two subclasses. That would reduce duplication.

File: source/common/http/filter_manager.h
Comment (kyessenov) [inline source/common/http/filter_manager.h:55]: Style: `filter_config_name_`.

File: test/common/http/filter_manager_test.cc
Comment (kyessenov) [inline test/common/http/filter_manager_test.cc:158]: Might be a good idea to name config and filter differently throughout, not obvious what this struct is.

File: test/common/http/filter_manager_test.cc
Comment (kyessenov) [inline test/common/http/filter_manager_test.cc:178]: Can you put these checks into a helper method in the test class? It's fairly repetitive.

**Review by kyessenov — CHANGES_REQUESTED** [review]
Code mostly LGTM, but there are a bunch of improvements to the style we can make.

File: source/common/http/filter_manager.cc
Comment (juanmolle) [reply source/common/http/filter_manager.cc:485]: understand.

Comment (juanmolle) [reply source/common/http/filter_manager.cc]: it is not possible to move the complete method to the base class because the virtual method comes from other base to the concrete class. I was not sure to use the same name, I didn't find a better one. please let me know if I should rename it

File: source/common/http/filter_manager.cc
Comment (kyessenov) [inline source/common/http/filter_manager.cc:375]: nit: unnecessary whitespace empty lines - the style in Envoy is not to put an empty line as a first line in the body

File: source/common/http/filter_manager.cc
Comment (kyessenov) [inline source/common/http/filter_manager.cc:499]: nit: same here, extra white space

File: test/common/http/filter_manager_test.cc
Comment (kyessenov) [inline test/common/http/filter_manager_test.cc:62]: nit: expectedName -> expected_name per convention

**Review by kyessenov — APPROVED** [review]
A couple of minor changes, but LGTM! Thanks for making improvements.

This needs a sign-off from a senior maintainer since it touches the core HTTP code.
/assign-from @envoyproxy/senior-maintainers

File: test/common/http/filter_manager_test.cc
Comment (wbpcode) [inline test/common/http/filter_manager_test.cc:30]: Please just move this `using` to the most inner namespace.

**Review by wbpcode — COMMENTED** [review]
LGTM overall. Some minor comments are added.

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**
